File: apps/macos/Cybara/Sources/Cybara/PetGame.swift
import SwiftUI

enum PetCareAction {
    case feed
    case play
    case rest
}

enum PetStage: String {
    case egg
    case hatching
    case baby
    case adult
}

enum PetMood {
    case happy
    case content
    case hungry
    case sleepy
    case bored
}

struct PetGameState: Codable, Equatable {
    var hunger: Int
    var energy: Int
    var joy: Int
    var bond: Int
    var cares: Int
    var updatedAt: Double
}

enum PetGame {
    static let statMax = 100
    static let hungerDecayPerMinute = 1.6
    static let energyDecayPerMinute = 1.1
    static let joyDecayPerMinute = 1.3
    static let maxOfflineMinutes = 12.0 * 60.0
    static let hatchCares = 3
    static let growCares = 8
    static let easterEggTaps = 5
    static let easterEggGapMs = 900.0
    static let storageKey = "cybara.pet.game"

    static func now() -> Double {
        Date().timeIntervalSince1970 * 1000
    }

    static func initialState(at moment: Double = PetGame.now()) -> PetGameState {
        PetGameState(hunger: 80, energy: 80, joy: 80, bond: 0, cares: 0, updatedAt: moment)
    }

    static func clamp(_ value: Double) -> Int {
        guard value.isFinite else { return 0 }
        return max(0, min(statMax, Int(value.rounded())))
    }

    static func decayed(_ state: PetGameState, at moment: Double) -> PetGameState {
        let elapsed = min(maxOfflineMinutes, max(0, (moment - state.updatedAt) / 60_000))
        guard elapsed > 0 else {
            var unchanged = state
            unchanged.updatedAt = moment
            return unchanged
        }
        return PetGameState(
            hunger: clamp(Double(state.hunger) - elapsed * hungerDecayPerMinute),
            energy: clamp(Double(state.energy) - elapsed * energyDecayPerMinute),
            joy: clamp(Double(state.joy) - elapsed * joyDecayPerMinute),
            bond: clamp(Double(state.bond)),
            cares: state.cares,
            updatedAt: moment
        )
    }

    static func applying(
        _ action: PetCareAction,
        to state: PetGameState,
        at moment: Double = PetGame.now()
    ) -> PetGameState {
        let current = decayed(state, at: moment)
        var next = current
        switch action {
        case .feed:
            next.hunger = clamp(Double(current.hunger) + 34)
            next.joy = clamp(Double(current.joy) + 6)
        case .play:
            next.joy = clamp(Double(current.joy) + 30)
            next.energy = clamp(Double(current.energy) - 12)
            next.hunger = clamp(Double(current.hunger) - 6)
        case .rest:
            next.energy = clamp(Double(current.energy) + 36)
            next.joy = clamp(Double(current.joy) - 4)
        }
        let cared = next.hunger > current.hunger
            || next.joy > current.joy
            || next.energy > current.energy
        next.bond = clamp(Double(current.bond) + (cared ? 2 : 0))
        next.cares = current.cares + 1
        next.updatedAt = moment
        return next
    }

    static func stage(of state: PetGameState) -> PetStage {
        if state.cares >= growCares { return .adult }
        if state.cares >= hatchCares { return .baby }
        if state.cares >= hatchCares - 1 { return .hatching }
        return .egg
    }

    static func mood(of state: PetGameState) -> PetMood {
        if state.hunger <= 30 { return .hungry }
        if state.energy <= 30 { return .sleepy }
        if state.joy <= 30 { return .bored }
        if state.hunger >= 70 && state.energy >= 70 && state.joy >= 70 { return .happy }
        return .content
    }

    static func moodLabel(_ mood: PetMood) -> String {
        switch mood {
        case .hungry: return "Wants a snack"
        case .sleepy: return "Getting sleepy"
        case .bored: return "Wants to play"
        case .happy: return "Thriving"
        case .content: return "Doing fine"
        }
    }

    static func level(of state: PetGameState) -> Int {
        1 + state.bond / 20
    }

    static func registerTap(
        taps: Int,
        lastTapAt: Double,
        at moment: Double
    ) -> (taps: Int, unlocked: Bool) {
        let next = moment - lastTapAt <= easterEggGapMs ? taps + 1 : 1
        let unlocked = next >= easterEggTaps
        return (unlocked ? 0 : next, unlocked)
    }

    static func load(from defaults: UserDefaults = .standard) -> PetGameState {
        guard
            let data = defaults.string(forKey: storageKey)?.data(using: .utf8),
            let stored = try? JSONDecoder().decode(PetGameState.self, from: data)
        else {
            return initialState()
        }
        return decayed(stored, at: now())
    }

    static func save(_ state: PetGameState, to defaults: UserDefaults = .standard) {
        guard
            let data = try? JSONEncoder().encode(state),
            let text = String(data: data, encoding: .utf8)
        else { return }
        defaults.set(text, forKey: storageKey)
    }
}

enum PetSprite {
    static let width = 16

    static let palette: [Character: Color] = [
        "o": Color(red: 0.231, green: 0.141, blue: 0.086),
        "f": Color(red: 0.776, green: 0.541, blue: 0.322),
        "d": Color(red: 0.549, green: 0.353, blue: 0.180),
        "m": Color(red: 0.604, green: 0.639, blue: 0.678),
        "e": Color(red: 0.141, green: 0.075, blue: 0.035),
        "w": Color.white,
        "r": Color(red: 0.961, green: 0.510, blue: 0.125),
        "c": Color(red: 0.949, green: 0.894, blue: 0.808),
        "l": Color(red: 0.875, green: 0.682, blue: 0.471),
        "s": Color(red: 0.659, green: 0.439, blue: 0.220),
    ]

    static let egg = [
        "................",
        "......oooo......",
        ".....occcco.....",
        "....occcccco....",
        "...occcccccco...",
        "...occcccccco...",
        "..occcccccccco..",
        "..orrccccccrro..",
        "..occcccccccco..",
        "..orrccccccrro..",
        "..occcccccccco..",
        "...occcccccco...",
        "....occcccco....",
        ".....occcco.....",
        "......oooo......",
        "................",
    ]

    static let eggCrack = [
        "................",
        "......oooo......",
        ".....occcco.....",
        "....occcccco....",
        "...occcccccco...",
        "...occoccocco...",
        "..occcoccoccco..",
        "..orrccccccrro..",
        "..occoccccocco..",
        "..orrccccccrro..",
        "..occcccccccco..",
        "...occcccccco...",
        "....occcccco....",
        ".....occcco.....",
        "......oooo......",
        "................",
    ]

    static let baby = [
        "................",
        "................",
        "................",
        "........r.......",
        ".......oro......",
        ".....oooooo.....",
        "....offffffo....",
        "....offffffo....",
        "....owffffwo....",
        "....offffffo....",
        "....offleffo....",
        "....offssffo....",
        ".....oooooo.....",
        ".....oo..oo.....",
        "................",
        "................",
    ]

    static let adult = [
        "................",
        ".......rr.......",
        "......orro......",
        "...oooooooooo...",
        ".ooffffffffffoo.",
        ".offffffffffffo.",
        ".offffffffffffo.",
        ".offweffffweffo.",
        ".offeeffffeeffo.",
        ".offffffffffmmo.",
        ".offflleellfffo.",
        ".offfllllllfffo.",
        ".offffssssssffo.",
        "..offsssssssfo..",
        "...oooooooooo...",
        "...oo......oo...",
    ]

    static let adultBlink: [String] = adult.enumerated().map { index, row in
        guard index == 7 || index == 8 else { return row }
        return row
            .replacingOccurrences(of: "we", with: "ff")
            .replacingOccurrences(of: "ee", with: "ff")
    }

    static let adultSleep: [String] = adult.enumerated().map { index, row in
        if index == 7 { return row.replacingOccurrences(of: "we", with: "oo") }
        if index == 8 { return row.replacingOccurrences(of: "ee", with: "ff") }
        return row
    }

    static func rows(for stage: PetStage, mood: PetMood, blink: Bool) -> [String] {
        switch stage {
        case .egg: return egg
        case .hatching: return eggCrack
        case .baby: return baby
        case .adult:
            if mood == .sleepy { return adultSleep }
            return blink ? adultBlink : adult
        }
    }
}

struct PetSpriteView: View {
    let rows: [String]
    let scale: CGFloat

    var body: some View {
        Canvas { context, _ in
            for (y, row) in rows.enumerated() {
                for (x, character) in row.enumerated() {
                    guard let color = PetSprite.palette[character] else { continue }
                    context.fill(
                        Path(
                            CGRect(
                                x: CGFloat(x) * scale,
                                y: CGFloat(y) * scale,
                                width: scale,
                                height: scale
                            )
                        ),
                        with: .color(color)
                    )
                }
            }
        }
        .frame(
            width: CGFloat(PetSprite.width) * scale,
            height: CGFloat(rows.count) * scale
        )
        .accessibilityLabel("Pixel art Cybara")
    }
}

File: shared/pet-game.ts
export type PetCareAction = "feed" | "play" | "rest";
export type PetMood = "happy" | "content" | "hungry" | "sleepy" | "bored";

export interface PetGameState {
  hunger: number;
  energy: number;
  joy: number;
  bond: number;
  cares: number;
  updatedAt: number;
}

export const PET_STAT_MAX = 100;
export const PET_DECAY_PER_MINUTE = { hunger: 1.6, energy: 1.1, joy: 1.3 };
const PET_MAX_OFFLINE_MINUTES = 12 * 60;

export const PET_EASTER_EGG_TAPS = 5;
export const PET_EASTER_EGG_GAP_MS = 900;

export function registerPetTap(
  taps: number,
  lastTapAt: number,
  now: number
): { taps: number; unlocked: boolean } {
  const next = now - lastTapAt <= PET_EASTER_EGG_GAP_MS ? taps + 1 : 1;
  return { taps: next >= PET_EASTER_EGG_TAPS ? 0 : next, unlocked: next >= PET_EASTER_EGG_TAPS };
}

export function createPetGameState(now = Date.now()): PetGameState {
  return { hunger: 80, energy: 80, joy: 80, bond: 0, cares: 0, updatedAt: now };
}

function clampStat(value: number): number {
  if (!Number.isFinite(value)) return 0;
  return Math.max(0, Math.min(PET_STAT_MAX, Math.round(value)));
}

export function decayPetGameState(state: PetGameState, now: number): PetGameState {
  const elapsedMinutes = Math.min(
    PET_MAX_OFFLINE_MINUTES,
    Math.max(0, (now - state.updatedAt) / 60_000)
  );
  if (elapsedMinutes <= 0) return { ...state, updatedAt: now };
  return {
    hunger: clampStat(state.hunger - elapsedMinutes * PET_DECAY_PER_MINUTE.hunger),
    energy: clampStat(state.energy - elapsedMinutes * PET_DECAY_PER_MINUTE.energy),
    joy: clampStat(state.joy - elapsedMinutes * PET_DECAY_PER_MINUTE.joy),
    bond: clampStat(state.bond),
    cares: state.cares,
    updatedAt: now,
  };
}

export function applyPetCareAction(
  state: PetGameState,
  action: PetCareAction,
  now = Date.now()
): PetGameState {
  const current = decayPetGameState(state, now);
  const next = { ...current };
  if (action === "feed") {
    next.hunger = clampStat(current.hunger + 34);
    next.joy = clampStat(current.joy + 6);
  } else if (action === "play") {
    next.joy = clampStat(current.joy + 30);
    next.energy = clampStat(current.energy - 12);
    next.hunger = clampStat(current.hunger - 6);
  } else {
    next.energy = clampStat(current.energy + 36);
    next.joy = clampStat(current.joy - 4);
  }
  const cared = next.hunger > current.hunger || next.joy > current.joy || next.energy > current.energy;
  next.bond = clampStat(current.bond + (cared ? 2 : 0));
  next.cares = current.cares + 1;
  next.updatedAt = now;
  return next;
}

export function petMood(state: PetGameState): PetMood {
  if (state.hunger <= 30) return "hungry";
  if (state.energy <= 30) return "sleepy";
  if (state.joy <= 30) return "bored";
  if (state.hunger >= 70 && state.energy >= 70 && state.joy >= 70) return "happy";
  return "content";
}

export function petMoodLabel(mood: PetMood): string {
  if (mood === "hungry") return "Wants a snack";
  if (mood === "sleepy") return "Getting sleepy";
  if (mood === "bored") return "Wants to play";
  if (mood === "happy") return "Thriving";
  return "Doing fine";
}

export const PET_HATCH_CARES = 3;
export const PET_GROW_CARES = 8;

export function petStage(state: PetGameState): "egg" | "hatching" | "baby" | "adult" {
  if (state.cares >= PET_GROW_CARES) return "adult";
  if (state.cares >= PET_HATCH_CARES) return "baby";
  if (state.cares >= PET_HATCH_CARES - 1) return "hatching";
  return "egg";
}

export function petLevel(state: PetGameState): number {
  return 1 + Math.floor(state.bond / 20);
}

export function resetPetGameState(now = Date.now()): PetGameState {
  return createPetGameState(now);
}

export function serializePetGameState(state: PetGameState): string {
  return JSON.stringify(state);
}

export function parsePetGameState(raw: string | null, now = Date.now()): PetGameState {
  if (!raw) return createPetGameState(now);
  try {
    const parsed = JSON.parse(raw) as Partial<PetGameState>;
    if (!parsed || typeof parsed !== "object") return createPetGameState(now);
    return decayPetGameState(
      {
        hunger: clampStat(Number(parsed.hunger ?? 80)),
        energy: clampStat(Number(parsed.energy ?? 80)),
        joy: clampStat(Number(parsed.joy ?? 80)),
        bond: clampStat(Number(parsed.bond ?? 0)),
        cares: Math.max(0, Math.round(Number(parsed.cares ?? 0)) || 0),
        updatedAt: Number.isFinite(Number(parsed.updatedAt)) ? Number(parsed.updatedAt) : now,
      },
      now
    );
  } catch {
    return createPetGameState(now);
  }
}
